tests: investigate testdriver race

Two recent failed test:

* https://github.com/govim/govim/runs/3542852476?check_suite_focus=true
* https://github.com/govim/govim/runs/3565621327?check_suite_focus=true

Example:
```
WARNING: DATA RACE
Read at 0x000000934e50 by goroutine 23:
  github.com/govim/govim/testdriver.NewTestDriver()
      /home/runner/govim/testdriver/testdriver.go:220 +0x147d
  github.com/govim/govim_test.TestScripts.func1.1()
      /home/runner/govim/govim_test.go:123 +0xc24
  github.com/rogpeppe/go-internal/testscript.(*TestScript).setup()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:347 +0x144c
  github.com/rogpeppe/go-internal/testscript.(*TestScript).run()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:402 +0x12c
  github.com/rogpeppe/go-internal/testscript.RunT.func1()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:247 +0x499
  github.com/rogpeppe/go-internal/testscript.tshim.Run.func1()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:183 +0x52
  testing.tRunner()
      /go/src/testing/testing.go:1123 +0x202

Previous write at 0x000000934e50 by goroutine 12:
  github.com/govim/govim/testdriver.NewTestDriver()
      /home/runner/govim/testdriver/testdriver.go:227 +0x14f8
  github.com/govim/govim_test.TestScripts.func1.1()
      /home/runner/govim/govim_test.go:123 +0xc24
  github.com/rogpeppe/go-internal/testscript.(*TestScript).setup()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:347 +0x144c
  github.com/rogpeppe/go-internal/testscript.(*TestScript).run()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:402 +0x12c
  github.com/rogpeppe/go-internal/testscript.RunT.func1()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:247 +0x499
  github.com/rogpeppe/go-internal/testscript.tshim.Run.func1()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:183 +0x52
  testing.tRunner()
      /go/src/testing/testing.go:1123 +0x202

Goroutine 23 (running) created at:
  testing.(*T).Run()
      /go/src/testing/testing.go:1168 +0x5bb
  github.com/rogpeppe/go-internal/testscript.tshim.Run()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:182 +0xb0
  github.com/rogpeppe/go-internal/testscript.RunT()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:223 +0x634
  github.com/rogpeppe/go-internal/testscript.Run()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:160 +0x384
  github.com/govim/govim_test.TestScripts.func1()
      /home/runner/govim/govim_test.go:50 +0x2d5
  testing.tRunner()
      /go/src/testing/testing.go:1123 +0x202

Goroutine 12 (running) created at:
  testing.(*T).Run()
      /go/src/testing/testing.go:1168 +0x5bb
  github.com/rogpeppe/go-internal/testscript.tshim.Run()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:182 +0xb0
  github.com/rogpeppe/go-internal/testscript.RunT()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:223 +0x634
  github.com/rogpeppe/go-internal/testscript.Run()
      /home/runner/gopath/pkg/mod/github.com/rogpeppe/go-internal@v1.8.0/testscript/testscript.go:160 +0x384
  github.com/govim/govim_test.TestScripts.func1()
      /home/runner/govim/govim_test.go:50 +0x2d5
  testing.tRunner()
      /go/src/testing/testing.go:1123 +0x202
==================
--- FAIL: TestScripts (0.00s)
    --- FAIL: TestScripts/scripts (0.00s)
        --- FAIL: TestScripts/scripts/bad_function (0.46s)
            testscript.go:397: 
                WORK=/artefacts/govim-race/script-bad_function
                # Test the handling of functions that return errors
                # call a bad function (0.094s)
                PASS
                
            testing.go:1038: race detected during execution of test
        --- FAIL: TestScripts/scripts/unscheduled (0.92s)
            testscript.go:397: 
                WORK=/artefacts/govim-race/script-unscheduled
                # Test that unscheduled callbacks can schedule work (0.557s)
                PASS
                
            testing.go:1038: race detected during execution of test
FAIL
```

--- a/cmd/govim/main_test.go
+++ b/cmd/govim/main_test.go
@@ -42,6 +42,10 @@ func init() {
 }
 
 func TestMain(m *testing.M) {
+	// Snapshot PATH since it is prepended by testscript.
+	if os.Getenv(testsetup.EnvTestPathEnv) == "" {
+		os.Setenv(testsetup.EnvTestPathEnv, os.Getenv("PATH"))
+	}
 	os.Exit(testscript.RunMain(m, map[string]func() int{
 		"vim":         testdriver.Vim,
 		"vimexprwait": testdriver.VimExprWait,
@@ -287,6 +291,7 @@ func TestInstallScripts(t *testing.T) {
 					"PLUGIN_PATH="+govimPath,
 					"GOPATH="+gopath,
 					"GOCACHE="+gocache,
+					testsetup.EnvTestPathEnv+"="+os.Getenv(testsetup.EnvTestPathEnv),
 					testsetup.EnvLoadTestAPI+"=true",
 				)
 				return nil
@@ -317,6 +322,7 @@ func TestInstallScripts(t *testing.T) {
 					"PLUGIN_PATH="+govimPath,
 					"GOPATH="+gopath,
 					"GOCACHE="+gocache,
+					testsetup.EnvTestPathEnv+"="+os.Getenv(testsetup.EnvTestPathEnv),
 					string(config.EnvVarUseGoplsFromPath)+"=true",
 					testsetup.EnvLoadTestAPI+"=true",
 				)
@@ -340,7 +346,7 @@ func execvim() int {
 	// testscript prepends PATH with the directory that contain
 	// subcommands. Since we are going to execute vim here, and also
 	// have a subcommand named "vim", we need to find the real vim path.
-	v, err := testsetup.RealVimPath()
+	v, err := testsetup.LookPath("vim")
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "failed to locate vim: %v", err)
 		return 1

--- a/govim_test.go
+++ b/govim_test.go
@@ -29,6 +29,8 @@ var (
 )
 
 func TestMain(m *testing.M) {
+	// Snapshot PATH since it is prepended by testscript.
+	os.Setenv(testsetup.EnvTestPathEnv, os.Getenv("PATH"))
 	os.Exit(testscript.RunMain(m, map[string]func() int{
 		"vim": testdriver.Vim,
 	}))

--- a/testdriver/testdriver.go
+++ b/testdriver/testdriver.go
@@ -196,7 +196,8 @@ func NewTestDriver(c *Config) (*TestDriver, error) {
 		"GOVIMTESTDRIVER_SOCKET="+res.driverListener.Addr().String(),
 	)
 
-	vimCmd := cmd
+	vimCmd := make([]string, len(cmd))
+	copy(vimCmd, cmd)
 	if e := os.Getenv("VIM_COMMAND"); e != "" {
 		vimCmd = strings.Fields(e)
 	}
@@ -220,7 +221,7 @@ func NewTestDriver(c *Config) (*TestDriver, error) {
 		if vimCmd[i] != "vim" {
 			continue
 		}
-		abs, err := testsetup.RealVimPath()
+		abs, err := testsetup.LookPath("vim")
 		if err != nil {
 			return nil, fmt.Errorf("failed to find vim binary: %v", err)
 		}
@@ -249,6 +250,7 @@ func NewTestDriver(c *Config) (*TestDriver, error) {
 func (d *TestDriver) Logf(format string, a ...interface{}) {
 	fmt.Fprintf(d.log, format+"\n", a...)
 }
+
 func (d *TestDriver) LogStripANSI(r io.Reader) {
 	scanner := bufio.NewScanner(r)
 	for {
@@ -588,7 +590,6 @@ func (d *TestDriver) listenDriver() error {
 		}
 		return nil
 	})
-
 	if err != nil {
 		return fmt.Errorf("%v", err)
 	}
@@ -933,8 +934,22 @@ func Condition(cond string) (bool, error) {
 }
 
 func getVimFlavourVersion(c testsetup.Command) (string, error) {
-	var allArgs []string
-	allArgs = append(allArgs, c...)
+	allArgs := make([]string, len(c))
+	// testscript prepends PATH with the directory that contain
+	// subcommands. Since we are going to execute vim here, and also
+	// also have a subcommand named "vim", we must specify the entire path if
+	// "vim" is a part of c.
+	for i := range c {
+		if c[i] != "vim" {
+			allArgs[i] = c[i]
+			continue
+		}
+		abs, err := testsetup.LookPath("vim")
+		if err != nil {
+			return "", fmt.Errorf("failed to find vim binary: %v", err)
+		}
+		allArgs[i] = abs
+	}
 	allArgs = append(allArgs, "-v", "--cmd", "echo v:versionlong | qall", "--not-a-term")
 	cmd := exec.Command(allArgs[0], allArgs[1:]...)
 	out, err := cmd.CombinedOutput()

--- a/testsetup/testsetup.go
+++ b/testsetup/testsetup.go
@@ -41,6 +41,11 @@ const (
 	// EnvTestRaceSlowdown is a floating point factor by which to adjust
 	// EnvErrLogMatchWait for race tests
 	EnvTestRaceSlowdown = "GOVIM_TEST_RACE_SLOWDOWN"
+
+	// EnvTestPathEnv is used to override PATH when looking for executables.
+	// Testscript prepends PATH with the subcommand folder, so EnvTestPathEnv
+	// acts as a snapshot of PATH before it was modified.
+	EnvTestPathEnv = "GOVIM_TEST_PATH_ENV"
 )
 
 // vim versions
@@ -157,25 +162,43 @@ func EnvLookupFlavorCommand() (flav govim.Flavor, cmd Command, err error) {
 	return flav, cmd, nil
 }
 
-// RealVimPath iterates over PATH returning the second vim
-// executable found. Reason being that testscript prepends
-// the PATH with the subcommand dir causing a conflict where
-// the first vim found is the subcommand.
-func RealVimPath() (path string, err error) {
-	var firstFound bool
-	for _, dir := range filepath.SplitList(os.Getenv("PATH")) {
-		path := filepath.Join(dir, "vim")
-		d, err := os.Stat(path)
+// LookPath searches for an executable named file, using EnvTestPathEnv as PATH
+// (if set). The ordinary PATH env is used if it isn't set.
+// If file contains a slash, it is tried directly.
+// The result may be an absolute path or a path relative to the current directory.
+func LookPath(file string) (string, error) {
+	findExecutable := func(file string) error {
+		d, err := os.Stat(file)
 		if err != nil {
-			continue
+			return err
 		}
-		if m := d.Mode(); m.IsDir() || m&0111 == 0 {
-			continue
+		if m := d.Mode(); !m.IsDir() && m&0111 != 0 {
+			return nil
 		}
-		if firstFound {
+		return os.ErrPermission
+	}
+
+	if strings.Contains(file, "/") {
+		err := findExecutable(file)
+		if err == nil {
+			return file, nil
+		}
+		return "", err
+	}
+
+	pathEnv := os.Getenv(EnvTestPathEnv)
+	if pathEnv == "" {
+		pathEnv = os.Getenv("PATH")
+	}
+	for _, dir := range filepath.SplitList(pathEnv) {
+		if dir == "" {
+			// Unix shell semantics: path element "" means "."
+			dir = "."
+		}
+		path := filepath.Join(dir, file)
+		if err := findExecutable(path); err == nil {
 			return path, nil
 		}
-		firstFound = true
 	}
-	return "", fmt.Errorf("not found")
+	return "", fmt.Errorf("failed to find executable in provided path")
 }

